packages/make-hardener/src/main.js

```diff
@@ -19,22 +19,41 @@
 // then copied from proposal-frozen-realms deep-freeze.js
 // then copied from SES/src/bundle/deepFreeze.js
 
+// @ts-check
+
 const { freeze, getOwnPropertyDescriptors, getPrototypeOf } = Object;
 const { ownKeys } = Reflect;
 
+/**
+ * @template T
+ * @typedef {(root: T) => T} Hardener
+ */
+
 /**
  * Create a `harden` function.
+ *
+ * @template T
+ * @returns {Hardener<T>}
  */
 function makeHardener() {
   const hardened = new WeakSet();
 
   const { harden } = {
+    /**
+     * @template T
+     * @param {T} root
+     * @returns {T}
+     */
     harden(root) {
       const toFreeze = new Set();
       const paths = new WeakMap();
 
       // If val is something we should be freezing but aren't yet,
       // add it to toFreeze.
+      /**
+       * @param {any} val
+       * @param {string} [path]
+       */
       function enqueue(val, path) {
```

Comment:
Huh, I'm surprised that `path` is optional but I see that it is. Since this PR is concerned about the typing of what's here, I'd simply change the actual function parameters to redundantly express this optionality without changing the meaning as follows
```suggestion
      /**
       * @param {any} val
       * @param {string} [path]
       */
      function enqueue(val, path = undefined) {
```

However, I wonder whether this is a mistake. Should the initial call be `enqueue(root, '')` rather than `enqueue(root)`? Feel free to either try that in this PR or we can postpone the question till after this PR.